File: sdk-typescript/test/iwfcompat/dead_end_flow.ts
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

import {
  Channel,
  StepList,
  StepMovement,
  deadEnd,
  doubleCodec,
  gracefulComplete,
  rpc,
  stringCodec,
  voidCodec,
  type Context,
  type Flow,
  type PersistenceSchema,
  type RPCResult,
  type Step,
  type StepDecision,
} from "../../src/index.js";

class DeadEndStep implements Step<void> {
  public readonly inputCodec = voidCodec;

  public getStepType(): string {
    return "DeadEndStep";
  }

  public execute(_context: Context, _input: void): StepDecision {
    return deadEnd();
  }
}

class DeadEndCompleteStep implements Step<void> {
  public readonly inputCodec = voidCodec;

  public getStepType(): string {
    return "DeadEndCompleteStep";
  }

  public execute(_context: Context, _input: void): StepDecision {
    return gracefulComplete();
  }
}

export class DeadEndFlow implements Flow {
  public readonly idleSignal = new Channel("idle-signal", voidCodec);
  public readonly idleInternal = new Channel("idle-internal", voidCodec);
  private readonly start = new DeadEndStep();
  private readonly complete = new DeadEndCompleteStep();

  public getFlowType(): string {
    return "DeadEndFlow";
  }

  public getSteps() {
    return StepList.startStep(this.start).otherSteps(this.complete);
  }

  public getPersistenceSchema(): PersistenceSchema {
    return { channels: [this.idleSignal, this.idleInternal] };
  }

  @rpc({ outputCodec: doubleCodec })
  public signalSize(context: Context): RPCResult<number> {
    return { output: this.idleSignal.size(context) };
  }

  @rpc({ outputCodec: doubleCodec })
  public publishInternal(context: Context): RPCResult<number> {
    this.idleInternal.publish(context, undefined);
    return { output: this.idleInternal.size(context) };
  }

  @rpc({ inputCodec: stringCodec, outputCodec: doubleCodec })
  public invoke(context: Context, _input: string): RPCResult<number> {
    if (context.flowId === "" || context.runId === "") {
      throw new Error("invalid RPC context");
    }
    return { output: 100, nextSteps: [StepMovement.of(this.complete, undefined)] };
  }
}

File: sdk-python/tests/iwfcompat/dead_end_flow.py
# Portions of this file are derived from indeedeng/iwf-java-sdk.
# Those portions are licensed under the Apache License, Version 2.0.
# See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
#
# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications are licensed under the Super Durable Source License 1.0.
# Third-Party Materials remain under the Apache License, Version 2.0.
# See LICENSE and LEGACY_NOTICES.md.

from dex import (
    Channel,
    Context,
    Flow,
    PersistenceSchema,
    RPCResult,
    Step,
    StepDecision,
    StepList,
    dead_end,
    rpc,
)


class DeadEndStep(Step[None]):
    def execute(self, context: Context, input: None) -> StepDecision:
        del context, input
        return dead_end()


class DeadEndFlow(Flow[None]):
    idle_signal = Channel[None]("idle-signal", type(None))

    def __init__(self) -> None:
        self.start = DeadEndStep()

    def get_steps(self) -> StepList[None]:
        return StepList.start_step(self.start)

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(self.idle_signal)

    @rpc
    def signal_size(self, context: Context) -> RPCResult[int]:
        return RPCResult(self.idle_signal.size(context))

    @rpc
    def publish_internal(self, context: Context) -> RPCResult[int]:
        self.idle_signal.publish(context, None)
        return RPCResult(self.idle_signal.size(context))
